includes/class-avatar-store.php

```diff
@@ -0,0 +1,196 @@
+<?php
+
+namespace Webmention;
+
+use WP_Comment;
+
+/**
+ * Avatar Store Class
+ *
+ */
+class Avatar_Store {
+	/**
+	 * Initialize the plugin, registering WordPress hooks.
+	 */
+	public static function init() {
+		// Store Avatars Locally
+		add_action( 'comment_post', array( static::class, 'store_avatar' ), 20 );
+		add_action( 'edit_comment', array( static::class, 'store_avatar' ), 20 );
+	}
+
+	/**
+	 * Return upload directory.
+	 *
+	 * @param string $filepath File Path. Optional
+	 * @param boolean $url Return a URL if true, otherwise the directory.
+	 * @return string URL of upload directory.
+	 */
+	public static function upload_directory( $filepath = '', $url = false ) {
+		$upload_dir  = wp_get_upload_dir();
+		$upload_dir  = $url ? $upload_dir['baseurl'] : $upload_dir['basedir'];
+		$upload_dir .= '/webmention/avatars/';
+		$upload_dir  = apply_filters( 'webmention_avatar_directory', $upload_dir, $url );
+		return $upload_dir . $filepath;
+	}
+
+	/**
+	 * Sideload Avatar
+	 *
+	 * @param string $url URL.
+	 * @param string $host Host.
+	 * @param string $author Author
+	 * @return string URL to Downloaded Image.
+	 *
+	 */
+	public static function sideload_avatar( $url, $host, $author ) {
+		// If the URL is inside the upload directory.
+		if ( str_contains( self::upload_directory( '', true ), $url ) ) {
+			return $url;
+		}
+
+		// Load dependencies.
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+		require_once ABSPATH . WPINC . '/media.php';
+
+		$filehandle = $host . '/' . md5( $author ) . '.jpg';
+		$filepath   = self::upload_directory( $filehandle );
+
+		// If this is a gravatar URL, automatically use gravatar to get the right size and file type.
+		if ( str_contains( 'gravatar.com', $url ) ) {
+			$hash    = wp_parse_url( $url, PHP_URL_PATH );
+			$default = get_option( 'avatar_default', 'mystery' );
+			$rating  = strtolower( get_option( 'avatar_rating' ) );
+			$hash    = str_replace( '/avatar/', '', $hash );
+			switch ( $default ) {
+				case 'mm':
+				case 'mystery':
+				case 'mysteryman':
+					$default = 'mp';
+					break;
+				case 'gravatar_default':
+					$default = false;
+					break;
+			}
+			$url  = 'https://www.gravatar.com/avatar/' . $hash . '.jpg';
+			$url  = add_query_arg(
+				array(
+					's' => WEBMENTION_AVATAR_SIZE,
+					'd' => $default, // Replace with our site default.
+					'r' => $rating,
+				),
+				$url
+			);
+			$file = download_url( $url, 300 );
+			if ( is_wp_error( $file ) ) {
+				return false;
+			}
+			@move_uploaded_file( $file, $filepath );
```

Comment:
please don't mute errors. what could possibly go wrong here?

Reply:
You wrote that line... but I'll see about updating it.

Reply:
is this my draft without changes? 😳

Reply:
This part is untouched from your original draft.

I mostly simplified the logic and the hooks.. but the storage and sideload code was as in your draft.

Reply:
Also, to be fair, the silencing of errors is done in the same way in WordPress core for the sideload functions.

Reply:
then let's keep it 🫤